api/server/routes_create_update.go

```diff
@@ -102,31 +131,32 @@ func (s *Server) ensureRoute(ctx context.Context, method string, wroute *models.
 }
 
 // ensureApp will only execute if it is on post or put. Patch is not allowed to create apps.
-func (s *Server) ensureApp(ctx context.Context, wroute *models.RouteWrapper, method string) error {
-	app, err := s.datastore.GetApp(ctx, wroute.Route.AppName)
+func (s *Server) ensureApp(ctx context.Context, appName string, wroute *models.RouteWrapper, method string) (string, error) {
+	appID, err := s.datastore.GetAppID(ctx, appName)
 	if err != nil && err != models.ErrAppsNotFound {
-		return err
-	} else if app == nil {
-		// Create a new application
-		newapp := &models.App{Name: wroute.Route.AppName}
+		return "", err
+	} else if appID == "" {
+		// Create a new application assuming that /:app/ is an app name
+		newapp := &models.App{Name: appName}
 
 		err = s.FireBeforeAppCreate(ctx, newapp)
 		if err != nil {
-			return err
+			return "", err
 		}
 
-		_, err = s.datastore.InsertApp(ctx, newapp)
+		app, err := s.datastore.InsertApp(ctx, newapp)
 		if err != nil {
-			return err
+			return "", err
 		}
+		appID = app.ID
 
-		err = s.FireAfterAppCreate(ctx, newapp)
+		err = s.FireAfterAppCreate(ctx, app)
 		if err != nil {
-			return err
+			return "", err
 		}
 
 	}
-	return nil
+	return appID, nil
```

Comment:
related note... i've got a long standing todo to put most of this logic into sql transactions as right now these actions are not atomic. the app id stuff didn't make this code any easier, may be worth pursuing pushing this logic into transactions sooner than later